types/eslint/index.d.ts

```diff
@@ -0,0 +1,386 @@
+    getTokensBetween(left: AstNode | Token | Comment, right: AstNode | Token | Comment, padding: FilterPredicate | CursorWithCountOptions): Token[];
+
+    commentsExistBetween(left: AstNode, right: AstNode): boolean;
+
+    getCommentsBefore(nodeOrToken: AstNode | Token): Comment[];
+
+    getCommentsAfter(nodeOrToken: AstNode | Token): Comment[];
+
+    getCommentsInside(node: AstNode): Comment[];
+}
+
+export interface SourceCodeConfig {
+    text: string;
+    ast: Ast;
+    parserServices?: ParserServices;
+    scopeManager?: ScopeManager;
+    visitorKeys?: VisitorKeys;
+}
+
+export interface Location {
+    line: number;
+    column: number;
+}
+
+type ParserServices = any;
+type ScopeManager = any;
+type VisitorKeys = any;
+
+export class SourceCode extends TokenStore {
+    text: string;
+    ast: Ast;
+    lines: string[];
+    hasBOM: boolean;
+    parserServices: ParserServices;
+    scopeManager: ScopeManager;
+    visitorKeys: VisitorKeys;
+
+    constructor(text: string, ast: Ast);
+    constructor(config: SourceCodeConfig);
+
+    static splitLines(text: string): string[];
+
+    getText(node?: AstNode, beforeCount?: number, afterCount?: number): string;
+
+    getLines(): string[];
+
+    getAllComments(): AstNode[];
+
+    getComments(node: AstNode): { leading: Comment[], trailing: Comment[] };
+
+    getJSDocComment(node: AstNode): Token | null;
+
+    getNodeByRangeIndex(index: number): AstNode | null;
+
+    isSpaceBetweenTokens(first: Token, second: Token): boolean;
+
+    getLocFromIndex(index: number): Location;
+
+    getIndexFromLoc(location: Location): number;
+}
+
+export type RuleLevel = 'off' | 'warn' | 'error' | 0 | 1 | 2;
+
+export interface Config {
+    rules?: {
+        [name: string]: (RuleLevel | [RuleLevel, any])
+    };
+    parser?: string;
+    parserOptions?: any;
+    settings?: any;
+    env?: { [name: string]: boolean };
+    globals?: { [name: string]: boolean };
+}
+
+export interface LintOptions {
+    filename?: string;
+    preprocess?: (code: string) => string[];
+    postprocess?: (problemLists: LintMessage[][]) => LintMessage[];
+    allowInlineConfig?: boolean;
+    reportUnusedDisableDirectives?: boolean;
+}
+
+export type Severity = 1 | 2;
+
+export interface LintMessage {
+    column: number;
+    line: number;
+    endColumn?: number;
+    endLine?: number;
+    ruleId: string | null;
+    message: string;
+    nodeType: string;
+    fatal?: true;
+    severity: Severity;
+    fix?: Fix;
+    source: string | null;
+}
+
+export interface FixOptions extends LintOptions {
+    fix?: boolean;
+}
+
+export type Range = [number, number];
+
+export interface Fix {
+    range: Range;
+    text: string;
+}
+
+export interface FixReport {
+    fixed: boolean;
+    output: string;
+    messages: LintMessage[];
+}
+
+export interface RuleModule {
+    create(context: RuleContext): any;
```

Comment:
```ts
interface RuleListener {
  onCodePathStart?(codePath: CodePath, node: ESTree.Node): void
  onCodePathEnd?(codePath: CodePath, node: ESTree.Node): void
  onCodePathSegmentStart?(segment: CodePath.Segment, node: ESTree.Node): void
  onCodePathSegmentEnd?(segment: CodePath.Segment, node: ESTree.Node): void
  onCodePathSegmentLoop?(fromSegment: CodePath.Segment, toSegment: CodePath.Segment, node: ESTree.Node): void
  [key: string]: (codePath: CodePath, node: ESTree.Node) => void | (segment: CodePath.Segment, node: ESTree.Node) => void | (fromSegment: CodePath.Segment, toSegment: CodePath.Segment, node: ESTree.Node) => void | (node: ESTree.Node) => void
}
```

Reply:
The indexed signature doesn't play well with the optional fields when `strictNullChecks` is enabled. See the [example](https://www.typescriptlang.org/play/index.html#src=type%20CodePath%20%3D%20'codePath'%3B%0D%0Atype%20AstNode%20%3D%20'node'%3B%0D%0A%0D%0Ainterface%20RuleListener%20%7B%0D%0A%20%20%20%20onCodePathStart%3F%3A%20(codePath%3A%20CodePath%2C%20node%3A%20AstNode)%20%3D%3E%20void%3B%0D%0A%20%20%20%20%5Bkey%3A%20string%5D%3A%20(codePath%3A%20CodePath%2C%20node%3A%20AstNode)%20%3D%3E%20void%3B%0D%0A%7D).

Let's keep it simple for now. We can revisit that in the future.

Reply:
[Counterexample](https://www.typescriptlang.org/play/index.html#src=type%20CodePath%20%3D%20'codePath'%3B%0D%0Atype%20AstNode%20%3D%20'node'%3B%0D%0A%0D%0Ainterface%20RuleListener%20%7B%0D%0A%20%20%20%20onCodePathStart%3F%3A%20(codePath%3A%20CodePath%2C%20node%3A%20AstNode)%20%3D%3E%20void%3B%0D%0A%20%20%20%20%5Bkey%3A%20string%5D%3A%20((codePath%3A%20CodePath%2C%20node%3A%20AstNode)%20%3D%3E%20void)%20%7C%20undefined%3B%0D%0A%7D) :)

Reply:
Ahah, nice!